### PR TITLE
Serverspec tests

### LINKED ISSUE #947: serverspec tests
This is a follow-up to #283. We should develop a set of serverspec tests (some initial work has been done in `spec-tests`) and use them to do TDD with our provisioning scripts.

### PR DESCRIPTION
**Pull request by conorsch**
Here's a pile of serverspec tests. This is a large PR, but the vast majority of changes are isolated in the `spec_tests` dir, and so shouldn't have any side-effects for the application code or provisioning process. The only files that have changed outside of the `spec_tests` dir are these:

```
.gitignore
Vagrantfile
docs/develop.md
docs/spec_tests.md
install_files/ansible-base/roles/app/templates/sites-available/document.conf
install_files/ansible-base/roles/restrict_direct_access_app/templates/app_rules_v4
install_files/ansible-base/roles/restrict_direct_access_mon/templates/mon_rules_v4
```

The changes to the Ansible files are strictly typographic, to make the serverspec tests more consistent. The change to the Vagrantfile is a conditional check to run the digitalocean provider call only if it's properly configured. See the changes to `docs/spec_tests.md` for how to run the new spectests, and make sure the recommended procedure works on your machine:

```
cd spec_tests
bundle update
bundle exec rake spec
```

The commands above will run `vagrant up` on the following VirtualBox machines:

* develop
* app-staging
* mon-staging
* build

There's a bit of magic in the `spec_tests/spec/spec_helper.rb` file to load variables for running the tests, but these variables are *not* read from the Ansible vars files. This separation is intentional, so that changes to Ansible files will require a manual change to the spectests in order to get them back in sync.

Further down the line these tests should be reorganized by roles, rather than hosts, but first we'll need to make that true of the Ansible layout, so the tests mirror the config being tested. More tests coming in the near future, but I consider the coverage provided by this PR sufficient to close #947.

### CODE REVIEW

**Comment by dolanjs** [comment]
lgtm. 
The changes to the non spec-test files don't break anything. And the spec tests are a great and needed addition.

File: Vagrantfile [hunk 1]
```diff
@@ -148,14 +148,19 @@ Vagrant.configure("2") do |config|
...
-    override.vm.box = 'digital_ocean'
-    override.vm.box_url = "https://github.com/smdahlen/vagrant-digitalocean/raw/master/box/digital_ocean.box"
-    provider.token = SNAP_API_TOKEN
-    provider.image = 'snapVagrantSSHkey'
-    provider.region = 'nyc2'
-    provider.size = '512mb'
+  # This is needed for the Snap-ci to provision the digital ocean vps.
+  # Check for presence of digitalocean vagrant plugin, and required env var,
+  # and only configure this provider if both conditions are met. Otherwise,
+  # even running `vagrant status` throws an error.
```
Comment (garrettr) [inline Vagrantfile]: :+1: I like this

**Comment by garrettr** [comment]
Nice job on the development guide for serverspec tests :+1:

**Comment by garrettr** [comment]
I need to finish testing the migration scripts, but once I'm done with that I will test this stuff on my Mac and document that setup in `docs/spec_tests.md`.

**Pull request merged**